src/ngx_http_lua_string.c

```diff
@@ -433,17 +433,24 @@ ngx_http_lua_ffi_unescape_uri(const u_char *src, size_t len, u_char *dst)
 
 
 size_t
-ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len)
+ngx_http_lua_ffi_uri_escaped_length(const u_char *src, size_t len,
+    int type)
 {
-    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len,
-                                             NGX_ESCAPE_URI_COMPONENT);
+    if (type < NGX_ESCAPE_URI || type > NGX_ESCAPE_MAIL_AUTH) {
+        return -1;
+    }
+    return len + 2 * ngx_http_lua_escape_uri(NULL, (u_char *) src, len, type);
 }
 
 
 void
-ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst)
+ngx_http_lua_ffi_escape_uri(const u_char *src, size_t len, u_char *dst,
+    int type)
 {
-    ngx_http_lua_escape_uri(dst, (u_char *) src, len, NGX_ESCAPE_URI_COMPONENT);
+    if (type < NGX_ESCAPE_URI || type > NGX_ESCAPE_MAIL_AUTH) {
```

Comment:
I think we only need to check it in the length function. Otherwise we are checking twice.